File: solutions/elixir/perfect-numbers/2/lib/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(number) when number <= 0,
    do: {:error, "Classification is only possible for natural numbers."}

  def classify(number) do
    aliquot_sum = number |> factors() |> Enum.sum()

    cond do
      aliquot_sum < number -> {:ok, :deficient}
      aliquot_sum > number -> {:ok, :abundant}
      true -> {:ok, :perfect}
    end
  end

  defp factors(n),
    do: for(factor <- Range.new(1, n - 1, 1), rem(n, factor) == 0, do: factor)
end

File: solutions/clojure/perfect-numbers/2/src/perfect_numbers.clj
(ns perfect-numbers)

(defn- aliquot-sum [n]
  (if (= n 1) 0
      (reduce
       (fn [acc factor]
         (if (zero? (rem n factor))
           (let [pair (/ n factor)]
             (+ acc factor (if (= factor pair) 0 pair)))
           acc))
       1
       (range 2 (inc (long (Math/sqrt n)))))))

(defn classify [n]
  (let [sum (aliquot-sum n)]
    (cond
      (< sum n) :deficient
      (> sum n) :abundant
      :else :perfect)))
